src/mococrw/mac.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+#include <memory>
+#include <vector>
+#include "openssl_wrap.h"
+
+namespace mococrw
+{
+
+/**
+ * @brief The MessageAuthenticationCode class is the abstract base class for the different implementation of
+ * message authentication codes.
+ */
+class MessageAuthenticationCode {
+public:
+
+    /**
+     * @brief ~MessageAuthenticationCode
+     */
+    virtual ~MessageAuthenticationCode();
+    /**
+     * Adds the message to the hash. This function may be invoked multiple times.
+     * For finishing the calculation of the MAC invoke finish().
+     *
+     * @param message chunk of data used for MAC.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    virtual void update(const std::vector<uint8_t>& message) = 0;
+
+    /**
+     * @brief Finalize the MAC.
+     *
+     * This function calculates the messaged authentication code.
+     *
+     * @throws MoCOCrWException if this function is invoked twice.
+     * @return The calculated message authentication code.
+     */
+    virtual std::vector<uint8_t> finish() = 0;
+
+    /**
+     * @brief Verifies the HMAC
+     *
+     * This function compares the given value (hmacValue), with the value calculated during finish().
+     * Invoke finish() before calling verify().
+     * Only the characters of the shorter value (hmacValue or calculated value) are checked for equality with the
```

Comment:
Do we really want this or should verify() just implicitly cause a finish()? I think this is more natural to use in the verification scenario.
Then we have
update()
update()
finish() for creating
and
update()
update()
verify()
for the verification

Reply:
Done